### PR TITLE
React app training middle ware

### PR DESCRIPTION
**Pull request by fededwolox**
## Summary
Adding MiddleWare

## TrelloCard
https://trello.com/c/hSZL727q/15-redux-ii-middleware-todo-finger-exercise

### CODE REVIEW

File: my-react-app/src/redux/login/actions.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import loginService from '../../services/loginService';
+
+export const actions = {
+  GET_TOKEN: 'GET_TOKEN',
+  GET_TOKEN_SUCCESS: 'GET_TOKEN_SUCCESS',
+  GET_TOKEN_FAILURE: 'GET_TOKEN_FAILURE'
+};
+
+const actionCreators = {
```
Comment (marianozicavo) [inline my-react-app/src/redux/login/actions.js:9]: You are not dispatching this actions anywhere.

File: my-react-app/src/redux/login/actions.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  login: credentials => async dispatch => {
+    dispatch({ type: actions.GET_TOKEN });
+    const response = await loginService.login(credentials);
+    if (response.ok) {
+      dispatch({
+        type: actions.GET_TOKEN_SUCCESS,
+        payload: response.data
+      });
+    } else {
+      if (response.data.emailInvalid) {
```
Comment (marianozicavo) [inline my-react-app/src/redux/login/actions.js]: Where did `emailInvalid` come from?

File: my-react-app/src/redux/login/actions.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        alert('Invalid email');
+      } else if (response.data.passwordInvalid) {
```
Comment (marianozicavo) [inline my-react-app/src/redux/login/actions.js]: Where did `passwordInvalid` come from?

File: my-react-app/src/redux/login/reducer.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+};
+
+export function reducerLogin(state = initialState, action) {
+  switch (action.type) {
+    case actions.GET_TOKEN:
+      return {
+        state
+      };
+    case actions.GET_TOKEN_SUCCESS:
+      localStorage.setItem('auth', action.payload.auth);
```
Comment (marianozicavo) [inline my-react-app/src/redux/login/reducer.js]: You shouldn't have side effects in a reducer, except for manipulating the redux store.

File: my-react-app/src/redux/login/reducer.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+      localStorage.setItem('token', action.payload.token);
+      return {
+        state,
```
Comment (marianozicavo) [inline my-react-app/src/redux/login/reducer.js]: I think you mean `...state`

File: my-react-app/src/redux/login/reducer.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        auth: true
+      };
+    case actions.GET_TOKEN_FAILURE:
+      localStorage.setItem('auth', action.payload.auth);
+      localStorage.setItem('token', '');
+      return {
+        state
+      };
+    default:
+      return { state };
```
Comment (marianozicavo) [inline my-react-app/src/redux/login/reducer.js]: You are englobing the whole state in an object each time this runs. If you want to keep the same state, `return state`

File: my-react-app/src/redux/login/reducer.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
+import { actions } from './actions';
+
+const initialState = {
+  token: localStorage.getItem('token'),
```
Comment (marianozicavo) [inline my-react-app/src/redux/login/reducer.js]: Let's just save the `token`(string), `email`(string) and `error`(boolean).

**Review by marianozicavo — CHANGES_REQUESTED** [review]

File: my-react-app/src/redux/login/actions.js (hunk 1 above)
Comment (fededwolox) [reply my-react-app/src/redux/login/actions.js]: I have dispatched them on screens/Login/index.js on the latest commit

File: my-react-app/src/app/screens/Login/index.js [hunk 8]
```diff
@@ -19,4 +21,9 @@ class LoginFormContainer extends Component {
   }
 }
 
-export default LoginFormContainer;
+const mapDispatchToProps = dispatch => ({
+  login: credentials =>
+    dispatch(actionCreators.login({ email: credentials.email, password: credentials.password }))
+});
+
+export default connect(mapDispatchToProps)(LoginFormContainer);
```
Comment (marianozicavo) [inline my-react-app/src/app/screens/Login/index.js]: The first parameter of connect is `mapStateToProps`. If you don't have that , do 
```
export default connect(null,mapDispatchToProps)(LoginFormContainer);
```

File: my-react-app/src/app/screens/Login/index.js [hunk 9]
```diff
@@ -19,4 +21,9 @@ class LoginFormContainer extends Component {
   }
 }
 
-export default LoginFormContainer;
+const mapDispatchToProps = dispatch => ({
+  login: credentials =>
```
Comment (marianozicavo) [inline my-react-app/src/app/screens/Login/index.js]: You are not using this prop anywhere, and it's missing in the propTypes

File: my-react-app/src/app/screens/Login/index.js [hunk 10]
```diff
@@ -19,4 +21,9 @@ class LoginFormContainer extends Component {
...
+    dispatch(actionCreators.login({ email: credentials.email, password: credentials.password }))
```
Comment (marianozicavo) [inline my-react-app/src/app/screens/Login/index.js]: `login: credentials => dispatch(actionCreators.login(credentials))`

File: my-react-app/src/redux/login/actions.js [hunk 11]
```diff
@@ -13,13 +13,11 @@ const actionCreators = {
     if (response.ok) {
       dispatch({
         type: actions.GET_TOKEN_SUCCESS,
-        payload: response.data
+        payload: { token: response.data, email: response.data }
```
Comment (marianozicavo) [inline my-react-app/src/redux/login/actions.js]: `response.data` is contains the token, not the email

File: my-react-app/src/redux/login/actions.js [hunk 12]
```diff
@@ -13,13 +13,11 @@ const actionCreators = {
...
       });
     } else {
-      if (response.data.emailInvalid) {
-        alert('Invalid email');
-      } else if (response.data.passwordInvalid) {
-        alert('Invalid password');
+      if (response.data.error) {
```
Comment (marianozicavo) [inline my-react-app/src/redux/login/actions.js]: This `if` is unnecesary

File: my-react-app/src/redux/login/reducer.js [hunk 13]
```diff
@@ -1,33 +1,28 @@
...
-  passwordInvalid: false
+  token: '',
+  email: '',
+  error: false
 };
 
 export function reducerLogin(state = initialState, action) {
   switch (action.type) {
     case actions.GET_TOKEN:
       return {
```
Comment (marianozicavo) [inline my-react-app/src/redux/login/reducer.js]: `return state`

File: my-react-app/src/redux/login/reducer.js [hunk 14]
```diff
@@ -1,33 +1,28 @@
...
-        state,
+        ...state,
         token: action.payload.token,
-        auth: true
+        email: action.payload.email
       };
     case actions.GET_TOKEN_FAILURE:
-      localStorage.setItem('auth', action.payload.auth);
-      localStorage.setItem('token', '');
       return {
```
Comment (marianozicavo) [inline my-react-app/src/redux/login/reducer.js:20]: Set `error` as `true` here

File: my-react-app/src/redux/login/actions.js (hunk 11 above)
Comment (marianozicavo) [inline my-react-app/src/redux/login/actions.js]: It should be `token: response.data.token`, as data is an object that has a `token` property

**Review by marianozicavo — CHANGES_REQUESTED** [review]

File: my-react-app/src/redux/login/reducer.js (hunk 14 above)
Comment (fededwolox) [reply my-react-app/src/redux/login/reducer.js:20]: it should be fixed tho

File: my-react-app/src/app/screens/Login/index.js [hunk 15]
```diff
@@ -1,29 +1,27 @@
...
 class LoginFormContainer extends Component {
-  handleSubmit = async values => {
-    const response = await loginService.login(values);
-    if (response.ok) {
-      setAuthorizationHeader(response.data.token);
-    } else {
-      alert('Invalid password');
-    }
-  };
+  handleSubmit = async values => this.props.login(values);
```
Comment (marianozicavo) [inline my-react-app/src/app/screens/Login/index.js]: You don't need this anymore. Just use `onSubmit={this.props.login}`

File: my-react-app/src/redux/login/actions.js [hunk 16]
```diff
@@ -13,12 +13,9 @@ const actionCreators = {
     if (response.ok) {
       dispatch({
         type: actions.GET_TOKEN_SUCCESS,
-        payload: { token: response.data, email: response.data }
+        payload: { token: response.data.token, email: credentials.email }
       });
     } else {
-      if (response.data.error) {
-        alert('Error');
```
Comment (marianozicavo) [inline my-react-app/src/redux/login/actions.js]: You should still use the alert, only remove the `if` that surrounds it

**Review by marianozicavo — CHANGES_REQUESTED** [review]

File: my-react-app/src/app/screens/Login/index.js [hunk 17]
```diff
@@ -7,14 +7,13 @@ import actionCreators from '../../../redux/login/actions.js';
 import LoginForm from './layout.js';
 
 class LoginFormContainer extends Component {
```
Comment (marianozicavo) [inline my-react-app/src/app/screens/Login/index.js]: Delete this `class` (and its proptypes) and we're done!

**Review by marianozicavo — CHANGES_REQUESTED** [review]

**Review by marianozicavo — APPROVED** [review]

**Pull request merged**